{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2024-026 Peak Annotator\n",
    "This notebook details methods to re-annotate peaks using ChIPseeker.\n",
    "\n",
    "## Initialize Environment\n",
    "\n",
    "Load libraries and set correct location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "ChIPseeker v1.38.0\n",
      "\n",
      "If you use ChIPseeker in published research, please cite:\n",
      "Qianwen Wang, Ming Li, Tianzhi Wu, Li Zhan, Lin Li, Meijun Chen, Wenqin Xie, Zijing Xie, Erqiang Hu, Shuangbin Xu, Guangchuang Yu. Exploring epigenomic datasets by ChIPseeker. Current Protocols 2022, 2(10): e585\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Load Libraries\n",
    "library(ChIPseeker)\n",
    "library(GenomicFeatures)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "\n",
    "# Load data\n",
    "exp_dir     <- \"/home/dalbao/2024-026-Tcf7ATAC\"\n",
    "nfcore_dir  <- \"01_nfcore_241228\"\n",
    "nb_dir      <- \"02_bp_notebooks\"\n",
    "prefix      <- \"peakAnno\"                                # bp for broad peak calling settings\n",
    "\n",
    "filter_file <- \"241231-bp-peakFilter.txt\"\n",
    "# Add folder structure to filter_file\n",
    "filter_file <- paste(nfcore_dir, filter_file, sep = \"/\")\n",
    "\n",
    "# nf-core ATAC-seq pipeline calls peaks from two different libraries:\n",
    "library     <- \"merged_library\"\n",
    "# If library == merged_library, lib_sh is mLb\n",
    "# else if it is == merged_replicates, lib_sh is mRp\n",
    "lib_sh      <- ifelse(library == \"merged_library\", \"mLb\", \"mRp\")\n",
    "\n",
    "# nf-core ATAC-seq pipeline calls peaks in two types, broad_peak and narrow_peak:\n",
    "peak_type   <- \"broad_peak\"\n",
    "# If peak_type == broad_peak, peak_sh is bp\n",
    "peak_sh     <- ifelse(peak_type == \"broad_peak\", \"bp\", \"np\")\n",
    "\n",
    "# Add peak type and date to prefix\n",
    "prefix <- paste0(format(Sys.Date(), \"%y%m%d\"), \"-\", peak_sh, \"-\", prefix)\n",
    "\n",
    "# Move to working directory\n",
    "setwd(exp_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import GTF Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Import genomic features from the file as a GRanges object ... \n",
      "OK\n",
      "\n",
      "Prepare the 'metadata' data frame ... \n",
      "OK\n",
      "\n",
      "Make the TxDb object ... \n",
      "Warning message in .get_cds_IDX(mcols0$type, mcols0$phase):\n",
      "“The \"phase\" metadata column contains non-NA values for features of type\n",
      "  stop_codon. This information was ignored.”\n",
      "OK\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Create a TxDb object for annotations \n",
    "txdb <- makeTxDbFromGFF(\"ref/genes.gtf\", format = \"gtf\") # GRCm38"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate Peaks with Flanks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">> loading peak file...\t\t\t\t 2025-01-13 06:16:03 AM \n",
      ">> preparing features information...\t\t 2025-01-13 06:16:04 AM \n",
      ">> identifying nearest features...\t\t 2025-01-13 06:16:04 AM \n",
      ">> calculating distance from peak to TSS...\t 2025-01-13 06:16:07 AM \n",
      ">> assigning genomic annotation...\t\t 2025-01-13 06:16:07 AM \n",
      ">> adding flank feature information from peaks...\t 2025-01-13 06:16:17 AM \n",
      ">> assigning chromosome lengths\t\t\t 2025-01-13 06:17:33 AM \n",
      ">> done...\t\t\t\t\t 2025-01-13 06:17:33 AM \n"
     ]
    }
   ],
   "source": [
    "# File name for unfiltered consensus peaks bed file\n",
    "consensus_file <- paste(\"consensus_peaks\", lib_sh, \"clN.bed\", sep = \".\")\n",
    "consensus_loc  <- paste(nfcore_dir,\n",
    "                        \"bwa\",\n",
    "                        library,\n",
    "                        \"macs2\",\n",
    "                        peak_type,\n",
    "                        \"consensus\",\n",
    "                        consensus_file,\n",
    "                        sep = \"/\")\n",
    "\n",
    "# Annotate peaks\n",
    "peak_anno <- annotatePeak(\n",
    "    peak = consensus_loc,\n",
    "    tssRegion = c(-3000, 3000),     # Region to consider as TSS proximal\n",
    "    TxDb = txdb,                    # Ensembl TxDb for GRCm38\n",
    "    addFlankGeneInfo = TRUE,        # Add gene information\n",
    "    flankDistance = 150000          # Distance to consider as gene body\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Annotated peaks generated by ChIPseeker\n",
       "285417/285417  peaks were annotated\n",
       "Genomic Annotation Summary:\n",
       "              Feature   Frequency\n",
       "9    Promoter (<=1kb)  9.65569675\n",
       "10   Promoter (1-2kb)  3.32706181\n",
       "11   Promoter (2-3kb)  3.11053651\n",
       "4              5' UTR  0.12052541\n",
       "3              3' UTR  1.42493264\n",
       "1            1st Exon  0.02908026\n",
       "7          Other Exon  2.77839092\n",
       "2          1st Intron  8.30644285\n",
       "8        Other Intron 21.07092430\n",
       "6  Downstream (<=300)  0.10475900\n",
       "5   Distal Intergenic 50.07164955"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check peak_anno\n",
    "peak_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 20</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>interval_id</th><th scope=col>V5</th><th scope=col>V6</th><th scope=col>annotation</th><th scope=col>geneChr</th><th scope=col>geneStart</th><th scope=col>geneEnd</th><th scope=col>geneLength</th><th scope=col>geneStrand</th><th scope=col>geneId</th><th scope=col>transcriptId</th><th scope=col>distanceToTSS</th><th scope=col>flank_txIds</th><th scope=col>flank_geneIds</th><th scope=col>flank_gene_distances</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>3008717</td><td>3009098</td><td>382</td><td>*</td><td>Interval_1</td><td>0</td><td>+</td><td>Distal Intergenic</td><td>29</td><td>3073253</td><td>3074322</td><td>1070</td><td>1</td><td>ENSMUSG00000102693</td><td>ENSMUST00000193812</td><td>-64155</td><td>ENSMUST00000193812;ENSMUST00000082908                                                         </td><td>ENSMUSG00000102693;ENSMUSG00000064842                                                         </td><td>-64155;-92918               </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>3026197</td><td>3027005</td><td>809</td><td>*</td><td>Interval_2</td><td>0</td><td>+</td><td>Distal Intergenic</td><td>29</td><td>3073253</td><td>3074322</td><td>1070</td><td>1</td><td>ENSMUSG00000102693</td><td>ENSMUST00000193812</td><td>-46248</td><td>ENSMUST00000193812;ENSMUST00000082908                                                         </td><td>ENSMUSG00000102693;ENSMUSG00000064842                                                         </td><td>-46248;-75011               </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>1</td><td>3062294</td><td>3062545</td><td>252</td><td>*</td><td>Interval_3</td><td>0</td><td>+</td><td>Distal Intergenic</td><td>29</td><td>3073253</td><td>3074322</td><td>1070</td><td>1</td><td>ENSMUSG00000102693</td><td>ENSMUST00000193812</td><td>-10708</td><td>ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                   </td><td>ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                   </td><td>-10708;-39471;153799;153087 </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>1</td><td>3062945</td><td>3063216</td><td>272</td><td>*</td><td>Interval_4</td><td>0</td><td>+</td><td>Distal Intergenic</td><td>29</td><td>3073253</td><td>3074322</td><td>1070</td><td>1</td><td>ENSMUSG00000102693</td><td>ENSMUST00000193812</td><td>-10037</td><td>ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                   </td><td>ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                   </td><td>-10037;-38800;153128;152416 </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1</td><td>3063384</td><td>3064024</td><td>641</td><td>*</td><td>Interval_5</td><td>0</td><td>+</td><td>Distal Intergenic</td><td>29</td><td>3073253</td><td>3074322</td><td>1070</td><td>1</td><td>ENSMUSG00000102693</td><td>ENSMUST00000193812</td><td> -9229</td><td>ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                   </td><td>ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                   </td><td>-9229;-37992;152320;151608  </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1</td><td>3084439</td><td>3085013</td><td>575</td><td>*</td><td>Interval_6</td><td>0</td><td>+</td><td>Distal Intergenic</td><td>29</td><td>3073253</td><td>3074322</td><td>1070</td><td>1</td><td>ENSMUSG00000102693</td><td>ENSMUST00000193812</td><td> 11186</td><td>ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265;ENSMUST00000070533</td><td>ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951;ENSMUSG00000051951</td><td>11186;-17003;131331;130619;0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 20\n",
       "\\begin{tabular}{r|llllllllllllllllllll}\n",
       "  & seqnames & start & end & width & strand & interval\\_id & V5 & V6 & annotation & geneChr & geneStart & geneEnd & geneLength & geneStrand & geneId & transcriptId & distanceToTSS & flank\\_txIds & flank\\_geneIds & flank\\_gene\\_distances\\\\\n",
       "  & <fct> & <int> & <int> & <int> & <fct> & <chr> & <int> & <chr> & <chr> & <int> & <int> & <int> & <int> & <int> & <chr> & <chr> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & 3008717 & 3009098 & 382 & * & Interval\\_1 & 0 & + & Distal Intergenic & 29 & 3073253 & 3074322 & 1070 & 1 & ENSMUSG00000102693 & ENSMUST00000193812 & -64155 & ENSMUST00000193812;ENSMUST00000082908                                                          & ENSMUSG00000102693;ENSMUSG00000064842                                                          & -64155;-92918               \\\\\n",
       "\t2 & 1 & 3026197 & 3027005 & 809 & * & Interval\\_2 & 0 & + & Distal Intergenic & 29 & 3073253 & 3074322 & 1070 & 1 & ENSMUSG00000102693 & ENSMUST00000193812 & -46248 & ENSMUST00000193812;ENSMUST00000082908                                                          & ENSMUSG00000102693;ENSMUSG00000064842                                                          & -46248;-75011               \\\\\n",
       "\t3 & 1 & 3062294 & 3062545 & 252 & * & Interval\\_3 & 0 & + & Distal Intergenic & 29 & 3073253 & 3074322 & 1070 & 1 & ENSMUSG00000102693 & ENSMUST00000193812 & -10708 & ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                    & ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                    & -10708;-39471;153799;153087 \\\\\n",
       "\t4 & 1 & 3062945 & 3063216 & 272 & * & Interval\\_4 & 0 & + & Distal Intergenic & 29 & 3073253 & 3074322 & 1070 & 1 & ENSMUSG00000102693 & ENSMUST00000193812 & -10037 & ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                    & ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                    & -10037;-38800;153128;152416 \\\\\n",
       "\t5 & 1 & 3063384 & 3064024 & 641 & * & Interval\\_5 & 0 & + & Distal Intergenic & 29 & 3073253 & 3074322 & 1070 & 1 & ENSMUSG00000102693 & ENSMUST00000193812 &  -9229 & ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                    & ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                    & -9229;-37992;152320;151608  \\\\\n",
       "\t6 & 1 & 3084439 & 3085013 & 575 & * & Interval\\_6 & 0 & + & Distal Intergenic & 29 & 3073253 & 3074322 & 1070 & 1 & ENSMUSG00000102693 & ENSMUST00000193812 &  11186 & ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265;ENSMUST00000070533 & ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951;ENSMUSG00000051951 & 11186;-17003;131331;130619;0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 20\n",
       "\n",
       "| <!--/--> | seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | interval_id &lt;chr&gt; | V5 &lt;int&gt; | V6 &lt;chr&gt; | annotation &lt;chr&gt; | geneChr &lt;int&gt; | geneStart &lt;int&gt; | geneEnd &lt;int&gt; | geneLength &lt;int&gt; | geneStrand &lt;int&gt; | geneId &lt;chr&gt; | transcriptId &lt;chr&gt; | distanceToTSS &lt;dbl&gt; | flank_txIds &lt;chr&gt; | flank_geneIds &lt;chr&gt; | flank_gene_distances &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | 3008717 | 3009098 | 382 | * | Interval_1 | 0 | + | Distal Intergenic | 29 | 3073253 | 3074322 | 1070 | 1 | ENSMUSG00000102693 | ENSMUST00000193812 | -64155 | ENSMUST00000193812;ENSMUST00000082908                                                          | ENSMUSG00000102693;ENSMUSG00000064842                                                          | -64155;-92918                |\n",
       "| 2 | 1 | 3026197 | 3027005 | 809 | * | Interval_2 | 0 | + | Distal Intergenic | 29 | 3073253 | 3074322 | 1070 | 1 | ENSMUSG00000102693 | ENSMUST00000193812 | -46248 | ENSMUST00000193812;ENSMUST00000082908                                                          | ENSMUSG00000102693;ENSMUSG00000064842                                                          | -46248;-75011                |\n",
       "| 3 | 1 | 3062294 | 3062545 | 252 | * | Interval_3 | 0 | + | Distal Intergenic | 29 | 3073253 | 3074322 | 1070 | 1 | ENSMUSG00000102693 | ENSMUST00000193812 | -10708 | ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                    | ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                    | -10708;-39471;153799;153087  |\n",
       "| 4 | 1 | 3062945 | 3063216 | 272 | * | Interval_4 | 0 | + | Distal Intergenic | 29 | 3073253 | 3074322 | 1070 | 1 | ENSMUSG00000102693 | ENSMUST00000193812 | -10037 | ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                    | ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                    | -10037;-38800;153128;152416  |\n",
       "| 5 | 1 | 3063384 | 3064024 | 641 | * | Interval_5 | 0 | + | Distal Intergenic | 29 | 3073253 | 3074322 | 1070 | 1 | ENSMUSG00000102693 | ENSMUST00000193812 |  -9229 | ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                    | ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                    | -9229;-37992;152320;151608   |\n",
       "| 6 | 1 | 3084439 | 3085013 | 575 | * | Interval_6 | 0 | + | Distal Intergenic | 29 | 3073253 | 3074322 | 1070 | 1 | ENSMUSG00000102693 | ENSMUST00000193812 |  11186 | ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265;ENSMUST00000070533 | ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951;ENSMUSG00000051951 | 11186;-17003;131331;130619;0 |\n",
       "\n"
      ],
      "text/plain": [
       "  seqnames start   end     width strand interval_id V5 V6 annotation       \n",
       "1 1        3008717 3009098 382   *      Interval_1  0  +  Distal Intergenic\n",
       "2 1        3026197 3027005 809   *      Interval_2  0  +  Distal Intergenic\n",
       "3 1        3062294 3062545 252   *      Interval_3  0  +  Distal Intergenic\n",
       "4 1        3062945 3063216 272   *      Interval_4  0  +  Distal Intergenic\n",
       "5 1        3063384 3064024 641   *      Interval_5  0  +  Distal Intergenic\n",
       "6 1        3084439 3085013 575   *      Interval_6  0  +  Distal Intergenic\n",
       "  geneChr geneStart geneEnd geneLength geneStrand geneId            \n",
       "1 29      3073253   3074322 1070       1          ENSMUSG00000102693\n",
       "2 29      3073253   3074322 1070       1          ENSMUSG00000102693\n",
       "3 29      3073253   3074322 1070       1          ENSMUSG00000102693\n",
       "4 29      3073253   3074322 1070       1          ENSMUSG00000102693\n",
       "5 29      3073253   3074322 1070       1          ENSMUSG00000102693\n",
       "6 29      3073253   3074322 1070       1          ENSMUSG00000102693\n",
       "  transcriptId       distanceToTSS\n",
       "1 ENSMUST00000193812 -64155       \n",
       "2 ENSMUST00000193812 -46248       \n",
       "3 ENSMUST00000193812 -10708       \n",
       "4 ENSMUST00000193812 -10037       \n",
       "5 ENSMUST00000193812  -9229       \n",
       "6 ENSMUST00000193812  11186       \n",
       "  flank_txIds                                                                                   \n",
       "1 ENSMUST00000193812;ENSMUST00000082908                                                         \n",
       "2 ENSMUST00000193812;ENSMUST00000082908                                                         \n",
       "3 ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                   \n",
       "4 ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                   \n",
       "5 ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265                   \n",
       "6 ENSMUST00000193812;ENSMUST00000082908;ENSMUST00000162897;ENSMUST00000159265;ENSMUST00000070533\n",
       "  flank_geneIds                                                                                 \n",
       "1 ENSMUSG00000102693;ENSMUSG00000064842                                                         \n",
       "2 ENSMUSG00000102693;ENSMUSG00000064842                                                         \n",
       "3 ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                   \n",
       "4 ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                   \n",
       "5 ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951                   \n",
       "6 ENSMUSG00000102693;ENSMUSG00000064842;ENSMUSG00000051951;ENSMUSG00000051951;ENSMUSG00000051951\n",
       "  flank_gene_distances        \n",
       "1 -64155;-92918               \n",
       "2 -46248;-75011               \n",
       "3 -10708;-39471;153799;153087 \n",
       "4 -10037;-38800;153128;152416 \n",
       "5 -9229;-37992;152320;151608  \n",
       "6 11186;-17003;131331;130619;0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract peak annotations\n",
    "peak_anno_df <- as.data.frame(peak_anno)\n",
    "\n",
    "# Rename the V4 column as interval_id using tidyr\n",
    "#peak_anno_df <- peak_anno_df %>% rename(V4= \"interval_id\")\n",
    "peak_anno_df <- peak_anno_df %>% rename(interval_id = \"V4\")\n",
    "head(peak_anno_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Kept Peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>interval_id</th><th scope=col>keep_peak</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>98</th><td>1</td><td>4496008</td><td>4497219</td><td>Interval_98 </td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>103</th><td>1</td><td>4747606</td><td>4748936</td><td>Interval_103</td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>105</th><td>1</td><td>4768351</td><td>4770866</td><td>Interval_105</td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>107</th><td>1</td><td>4779308</td><td>4780669</td><td>Interval_107</td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>108</th><td>1</td><td>4785113</td><td>4786365</td><td>Interval_108</td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>110</th><td>1</td><td>4807125</td><td>4808614</td><td>Interval_110</td><td>TRUE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & chr & start & end & interval\\_id & keep\\_peak\\\\\n",
       "  & <chr> & <int> & <int> & <chr> & <lgl>\\\\\n",
       "\\hline\n",
       "\t98 & 1 & 4496008 & 4497219 & Interval\\_98  & TRUE\\\\\n",
       "\t103 & 1 & 4747606 & 4748936 & Interval\\_103 & TRUE\\\\\n",
       "\t105 & 1 & 4768351 & 4770866 & Interval\\_105 & TRUE\\\\\n",
       "\t107 & 1 & 4779308 & 4780669 & Interval\\_107 & TRUE\\\\\n",
       "\t108 & 1 & 4785113 & 4786365 & Interval\\_108 & TRUE\\\\\n",
       "\t110 & 1 & 4807125 & 4808614 & Interval\\_110 & TRUE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | chr &lt;chr&gt; | start &lt;int&gt; | end &lt;int&gt; | interval_id &lt;chr&gt; | keep_peak &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 98 | 1 | 4496008 | 4497219 | Interval_98  | TRUE |\n",
       "| 103 | 1 | 4747606 | 4748936 | Interval_103 | TRUE |\n",
       "| 105 | 1 | 4768351 | 4770866 | Interval_105 | TRUE |\n",
       "| 107 | 1 | 4779308 | 4780669 | Interval_107 | TRUE |\n",
       "| 108 | 1 | 4785113 | 4786365 | Interval_108 | TRUE |\n",
       "| 110 | 1 | 4807125 | 4808614 | Interval_110 | TRUE |\n",
       "\n"
      ],
      "text/plain": [
       "    chr start   end     interval_id  keep_peak\n",
       "98  1   4496008 4497219 Interval_98  TRUE     \n",
       "103 1   4747606 4748936 Interval_103 TRUE     \n",
       "105 1   4768351 4770866 Interval_105 TRUE     \n",
       "107 1   4779308 4780669 Interval_107 TRUE     \n",
       "108 1   4785113 4786365 Interval_108 TRUE     \n",
       "110 1   4807125 4808614 Interval_110 TRUE     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>interval_id</th><th scope=col>annotation</th><th scope=col>geneChr</th><th scope=col>geneStart</th><th scope=col>geneEnd</th><th scope=col>geneLength</th><th scope=col>geneStrand</th><th scope=col>geneId</th><th scope=col>distanceToTSS</th><th scope=col>flank_geneIds</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>98</th><td>1</td><td>4496009</td><td>4497219</td><td>1211</td><td>*</td><td>Interval_98 </td><td>Promoter (&lt;=1kb) </td><td>29</td><td>4492458</td><td>4496330</td><td> 3873</td><td>2</td><td>ENSMUSG00000025902</td><td><span style=white-space:pre-wrap>     0</span></td><td><span style=white-space:pre-wrap>ENSMUSG00000025900;ENSMUSG00000104123;ENSMUSG00000025902;ENSMUSG00000104238;ENSMUSG00000102269;ENSMUSG00000096126;ENSMUSG00000103003;ENSMUSG00000104328;ENSMUSG00000102735                                                                                                                  </span></td></tr>\n",
       "\t<tr><th scope=row>103</th><td>1</td><td>4747607</td><td>4748936</td><td>1330</td><td>*</td><td>Interval_103</td><td>Distal Intergenic</td><td>29</td><td>4735046</td><td>4735676</td><td>  631</td><td>2</td><td>ENSMUSG00000103265</td><td>-11931</td><td>ENSMUSG00000102735;ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588                                                         </td></tr>\n",
       "\t<tr><th scope=row>105</th><td>1</td><td>4768352</td><td>4770866</td><td>2515</td><td>*</td><td>Interval_105</td><td>Promoter (&lt;=1kb) </td><td>29</td><td>4771131</td><td>4772199</td><td> 1069</td><td>1</td><td>ENSMUSG00000103922</td><td><span style=white-space:pre-wrap>  -265</span></td><td><span style=white-space:pre-wrap>ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459                                      </span></td></tr>\n",
       "\t<tr><th scope=row>107</th><td>1</td><td>4779309</td><td>4780669</td><td>1361</td><td>*</td><td>Interval_107</td><td>Promoter (&lt;=1kb) </td><td>29</td><td>4778063</td><td>4779212</td><td> 1150</td><td>2</td><td>ENSMUSG00000102275</td><td><span style=white-space:pre-wrap>   -97</span></td><td><span style=white-space:pre-wrap>ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305                   </span></td></tr>\n",
       "\t<tr><th scope=row>108</th><td>1</td><td>4785114</td><td>4786365</td><td>1252</td><td>*</td><td>Interval_108</td><td>Promoter (&lt;=1kb) </td><td>29</td><td>4783572</td><td>4785692</td><td> 2121</td><td>2</td><td>ENSMUSG00000033845</td><td><span style=white-space:pre-wrap>     0</span></td><td><span style=white-space:pre-wrap>ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305                   </span></td></tr>\n",
       "\t<tr><th scope=row>110</th><td>1</td><td>4807126</td><td>4808614</td><td>1489</td><td>*</td><td>Interval_110</td><td>Promoter (&lt;=1kb) </td><td>29</td><td>4807788</td><td>4848410</td><td>40623</td><td>1</td><td>ENSMUSG00000025903</td><td><span style=white-space:pre-wrap>     0</span></td><td>ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305;ENSMUSG00000102653</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & seqnames & start & end & width & strand & interval\\_id & annotation & geneChr & geneStart & geneEnd & geneLength & geneStrand & geneId & distanceToTSS & flank\\_geneIds\\\\\n",
       "  & <fct> & <int> & <int> & <int> & <fct> & <chr> & <chr> & <int> & <int> & <int> & <int> & <int> & <chr> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t98 & 1 & 4496009 & 4497219 & 1211 & * & Interval\\_98  & Promoter (<=1kb)  & 29 & 4492458 & 4496330 &  3873 & 2 & ENSMUSG00000025902 &      0 & ENSMUSG00000025900;ENSMUSG00000104123;ENSMUSG00000025902;ENSMUSG00000104238;ENSMUSG00000102269;ENSMUSG00000096126;ENSMUSG00000103003;ENSMUSG00000104328;ENSMUSG00000102735                                                                                                                  \\\\\n",
       "\t103 & 1 & 4747607 & 4748936 & 1330 & * & Interval\\_103 & Distal Intergenic & 29 & 4735046 & 4735676 &   631 & 2 & ENSMUSG00000103265 & -11931 & ENSMUSG00000102735;ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588                                                         \\\\\n",
       "\t105 & 1 & 4768352 & 4770866 & 2515 & * & Interval\\_105 & Promoter (<=1kb)  & 29 & 4771131 & 4772199 &  1069 & 1 & ENSMUSG00000103922 &   -265 & ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459                                      \\\\\n",
       "\t107 & 1 & 4779309 & 4780669 & 1361 & * & Interval\\_107 & Promoter (<=1kb)  & 29 & 4778063 & 4779212 &  1150 & 2 & ENSMUSG00000102275 &    -97 & ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305                   \\\\\n",
       "\t108 & 1 & 4785114 & 4786365 & 1252 & * & Interval\\_108 & Promoter (<=1kb)  & 29 & 4783572 & 4785692 &  2121 & 2 & ENSMUSG00000033845 &      0 & ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305                   \\\\\n",
       "\t110 & 1 & 4807126 & 4808614 & 1489 & * & Interval\\_110 & Promoter (<=1kb)  & 29 & 4807788 & 4848410 & 40623 & 1 & ENSMUSG00000025903 &      0 & ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305;ENSMUSG00000102653\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 15\n",
       "\n",
       "| <!--/--> | seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | interval_id &lt;chr&gt; | annotation &lt;chr&gt; | geneChr &lt;int&gt; | geneStart &lt;int&gt; | geneEnd &lt;int&gt; | geneLength &lt;int&gt; | geneStrand &lt;int&gt; | geneId &lt;chr&gt; | distanceToTSS &lt;dbl&gt; | flank_geneIds &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 98 | 1 | 4496009 | 4497219 | 1211 | * | Interval_98  | Promoter (&lt;=1kb)  | 29 | 4492458 | 4496330 |  3873 | 2 | ENSMUSG00000025902 |      0 | ENSMUSG00000025900;ENSMUSG00000104123;ENSMUSG00000025902;ENSMUSG00000104238;ENSMUSG00000102269;ENSMUSG00000096126;ENSMUSG00000103003;ENSMUSG00000104328;ENSMUSG00000102735                                                                                                                   |\n",
       "| 103 | 1 | 4747607 | 4748936 | 1330 | * | Interval_103 | Distal Intergenic | 29 | 4735046 | 4735676 |   631 | 2 | ENSMUSG00000103265 | -11931 | ENSMUSG00000102735;ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588                                                          |\n",
       "| 105 | 1 | 4768352 | 4770866 | 2515 | * | Interval_105 | Promoter (&lt;=1kb)  | 29 | 4771131 | 4772199 |  1069 | 1 | ENSMUSG00000103922 |   -265 | ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459                                       |\n",
       "| 107 | 1 | 4779309 | 4780669 | 1361 | * | Interval_107 | Promoter (&lt;=1kb)  | 29 | 4778063 | 4779212 |  1150 | 2 | ENSMUSG00000102275 |    -97 | ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305                    |\n",
       "| 108 | 1 | 4785114 | 4786365 | 1252 | * | Interval_108 | Promoter (&lt;=1kb)  | 29 | 4783572 | 4785692 |  2121 | 2 | ENSMUSG00000033845 |      0 | ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305                    |\n",
       "| 110 | 1 | 4807126 | 4808614 | 1489 | * | Interval_110 | Promoter (&lt;=1kb)  | 29 | 4807788 | 4848410 | 40623 | 1 | ENSMUSG00000025903 |      0 | ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305;ENSMUSG00000102653 |\n",
       "\n"
      ],
      "text/plain": [
       "    seqnames start   end     width strand interval_id  annotation       \n",
       "98  1        4496009 4497219 1211  *      Interval_98  Promoter (<=1kb) \n",
       "103 1        4747607 4748936 1330  *      Interval_103 Distal Intergenic\n",
       "105 1        4768352 4770866 2515  *      Interval_105 Promoter (<=1kb) \n",
       "107 1        4779309 4780669 1361  *      Interval_107 Promoter (<=1kb) \n",
       "108 1        4785114 4786365 1252  *      Interval_108 Promoter (<=1kb) \n",
       "110 1        4807126 4808614 1489  *      Interval_110 Promoter (<=1kb) \n",
       "    geneChr geneStart geneEnd geneLength geneStrand geneId            \n",
       "98  29      4492458   4496330  3873      2          ENSMUSG00000025902\n",
       "103 29      4735046   4735676   631      2          ENSMUSG00000103265\n",
       "105 29      4771131   4772199  1069      1          ENSMUSG00000103922\n",
       "107 29      4778063   4779212  1150      2          ENSMUSG00000102275\n",
       "108 29      4783572   4785692  2121      2          ENSMUSG00000033845\n",
       "110 29      4807788   4848410 40623      1          ENSMUSG00000025903\n",
       "    distanceToTSS\n",
       "98       0       \n",
       "103 -11931       \n",
       "105   -265       \n",
       "107    -97       \n",
       "108      0       \n",
       "110      0       \n",
       "    flank_geneIds                                                                                                                                                                                                                                                                               \n",
       "98  ENSMUSG00000025900;ENSMUSG00000104123;ENSMUSG00000025902;ENSMUSG00000104238;ENSMUSG00000102269;ENSMUSG00000096126;ENSMUSG00000103003;ENSMUSG00000104328;ENSMUSG00000102735                                                                                                                  \n",
       "103 ENSMUSG00000102735;ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588                                                         \n",
       "105 ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459                                      \n",
       "107 ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305                   \n",
       "108 ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305                   \n",
       "110 ENSMUSG00000098104;ENSMUSG00000102175;ENSMUSG00000088000;ENSMUSG00000103265;ENSMUSG00000103922;ENSMUSG00000033845;ENSMUSG00000102275;ENSMUSG00000025903;ENSMUSG00000104217;ENSMUSG00000033813;ENSMUSG00000062588;ENSMUSG00000103280;ENSMUSG00000002459;ENSMUSG00000091305;ENSMUSG00000102653"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Import peak filter\n",
    "kept_peaks <- read.table(filter_file, header = TRUE)\n",
    "kept_peaks <- kept_peaks[kept_peaks$keep == TRUE, ]\n",
    "head(kept_peaks)\n",
    "\n",
    "# Remove transcripts\n",
    "peak_anno_df <- peak_anno_df %>% select(-V5, -V6, -transcriptId, -flank_txIds, -flank_gene_distances)\n",
    "\n",
    "# Cleanup (fuck)\n",
    "geneIds <- strsplit(peak_anno_df$flank_geneIds, \";\")\n",
    "geneIds <- lapply(geneIds, unique)\n",
    "geneIds <- sapply(geneIds, paste, collapse = \";\")\n",
    "peak_anno_df$flank_geneIds <- geneIds\n",
    "\n",
    "fn <- paste(nfcore_dir, \"/\", prefix, \"-Unfiltered-NoSymbol.csv\", sep = \"\")\n",
    "# Save the peak kept peaks and annotated peaks\n",
    "write.csv(  peak_anno_df, file = fn,\n",
    "            quote = FALSE, row.names = FALSE)\n",
    "\n",
    "# Only keep peak_anno_df observations in kept_peaks$keep\n",
    "peak_anno_df <- peak_anno_df[peak_anno_df$interval_id %in% kept_peaks$interval_id, ]\n",
    "\n",
    "head(peak_anno_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert EnsemblIDs to Gene Symbols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GeneID</th><th scope=col>GeneSymbol</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSMUSG00000064372</td><td>mt-Tp  </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSMUSG00000064371</td><td>mt-Tt  </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENSMUSG00000064370</td><td>mt-Cytb</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENSMUSG00000064369</td><td>mt-Te  </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENSMUSG00000064368</td><td>mt-Nd6 </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENSMUSG00000064367</td><td>mt-Nd5 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & GeneID & GeneSymbol\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENSMUSG00000064372 & mt-Tp  \\\\\n",
       "\t2 & ENSMUSG00000064371 & mt-Tt  \\\\\n",
       "\t3 & ENSMUSG00000064370 & mt-Cytb\\\\\n",
       "\t4 & ENSMUSG00000064369 & mt-Te  \\\\\n",
       "\t5 & ENSMUSG00000064368 & mt-Nd6 \\\\\n",
       "\t6 & ENSMUSG00000064367 & mt-Nd5 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | GeneID &lt;chr&gt; | GeneSymbol &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | ENSMUSG00000064372 | mt-Tp   |\n",
       "| 2 | ENSMUSG00000064371 | mt-Tt   |\n",
       "| 3 | ENSMUSG00000064370 | mt-Cytb |\n",
       "| 4 | ENSMUSG00000064369 | mt-Te   |\n",
       "| 5 | ENSMUSG00000064368 | mt-Nd6  |\n",
       "| 6 | ENSMUSG00000064367 | mt-Nd5  |\n",
       "\n"
      ],
      "text/plain": [
       "  GeneID             GeneSymbol\n",
       "1 ENSMUSG00000064372 mt-Tp     \n",
       "2 ENSMUSG00000064371 mt-Tt     \n",
       "3 ENSMUSG00000064370 mt-Cytb   \n",
       "4 ENSMUSG00000064369 mt-Te     \n",
       "5 ENSMUSG00000064368 mt-Nd6    \n",
       "6 ENSMUSG00000064367 mt-Nd5    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load EnsemblID conversion\n",
    "ensembl2symbol <- read.table(\"ref/ensembl_102.txt\", header = TRUE, sep = \"\\t\", stringsAsFactors = FALSE)\n",
    "colnames(ensembl2symbol) <- c(\"GeneID\", \"GeneSymbol\")\n",
    "\n",
    "# Check df\n",
    "head(ensembl2symbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Function to convert EnsemblIDs to GeneSymbols\n",
    "convert_ids <- function(ids, conversion_df) {\n",
    "    sapply(ids, function(id) {\n",
    "        id_list <- unlist(strsplit(id, \";\"))\n",
    "        symbol_list <- sapply(id_list, function(single_id) {\n",
    "            symbol <- conversion_df$GeneSymbol[conversion_df$GeneID == single_id]\n",
    "            if (length(symbol) == 0 || is.na(symbol[1]) || symbol[1] == \"\") {\n",
    "                return(single_id)\n",
    "            } else {\n",
    "                return(symbol)\n",
    "            }\n",
    "        })\n",
    "        paste(symbol_list, collapse = \";\")\n",
    "    })\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Keep only items in ensemble2symbol that are in peak_anno_df\n",
    "uniqueids <- unique(unlist(strsplit(peak_anno_df$flank_geneIds, \";\")))\n",
    "uniqueids <- unique(c(uniqueids, peak_anno_df$flank_geneId))\n",
    "\n",
    "ensembl2symbol <- ensembl2symbol[ensembl2symbol$GeneID %in% uniqueids, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Done Annotation\"\n"
     ]
    }
   ],
   "source": [
    "# Convert geneId and flank_geneIds columns\n",
    "peak_anno_df$geneSymbol <- convert_ids(peak_anno_df$geneId, ensembl2symbol)\n",
    "peak_anno_df$flank_geneSymbols <- convert_ids(peak_anno_df$flank_geneIds, ensembl2symbol)\n",
    "print(\"Done Annotation\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 6 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>used</th><th scope=col>(Mb)</th><th scope=col>gc trigger</th><th scope=col>(Mb)</th><th scope=col>max used</th><th scope=col>(Mb)</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Ncells</th><td>11708379</td><td>625.3</td><td> 33603878</td><td>1794.7</td><td> 52506058</td><td>2804.2</td></tr>\n",
       "\t<tr><th scope=row>Vcells</th><td>46129721</td><td>352.0</td><td>135546944</td><td>1034.2</td><td>264740105</td><td>2019.9</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 6 of type dbl\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & used & (Mb) & gc trigger & (Mb) & max used & (Mb)\\\\\n",
       "\\hline\n",
       "\tNcells & 11708379 & 625.3 &  33603878 & 1794.7 &  52506058 & 2804.2\\\\\n",
       "\tVcells & 46129721 & 352.0 & 135546944 & 1034.2 & 264740105 & 2019.9\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 6 of type dbl\n",
       "\n",
       "| <!--/--> | used | (Mb) | gc trigger | (Mb) | max used | (Mb) |\n",
       "|---|---|---|---|---|---|---|\n",
       "| Ncells | 11708379 | 625.3 |  33603878 | 1794.7 |  52506058 | 2804.2 |\n",
       "| Vcells | 46129721 | 352.0 | 135546944 | 1034.2 | 264740105 | 2019.9 |\n",
       "\n"
      ],
      "text/plain": [
       "       used     (Mb)  gc trigger (Mb)   max used  (Mb)  \n",
       "Ncells 11708379 625.3  33603878  1794.7  52506058 2804.2\n",
       "Vcells 46129721 352.0 135546944  1034.2 264740105 2019.9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"01_nfcore_241228/250113-bp-peakAnno-Filtered.csv\"\n"
     ]
    }
   ],
   "source": [
    "# For some reason this is needed or else write times suffer.\n",
    "gc(full = TRUE)\n",
    "\n",
    "# Save annotated file\n",
    "fn <- paste(nfcore_dir, \"/\", prefix, \"-Filtered.csv\", sep = \"\")\n",
    "write.csv(  peak_anno_df, file = fn,\n",
    "            quote = TRUE, row.names = FALSE)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "test <- peak_anno_df[!grepl(\"ENS\", peak_anno_df$geneId), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 17</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>interval_id</th><th scope=col>annotation</th><th scope=col>geneChr</th><th scope=col>geneStart</th><th scope=col>geneEnd</th><th scope=col>geneLength</th><th scope=col>geneStrand</th><th scope=col>geneId</th><th scope=col>distanceToTSS</th><th scope=col>flank_geneIds</th><th scope=col>geneSymbol</th><th scope=col>flank_geneSymbols</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 17\n",
       "\\begin{tabular}{lllllllllllllllll}\n",
       " seqnames & start & end & width & strand & interval\\_id & annotation & geneChr & geneStart & geneEnd & geneLength & geneStrand & geneId & distanceToTSS & flank\\_geneIds & geneSymbol & flank\\_geneSymbols\\\\\n",
       " <fct> & <int> & <int> & <int> & <fct> & <chr> & <chr> & <int> & <int> & <int> & <int> & <int> & <chr> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 17\n",
       "\n",
       "| seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | interval_id &lt;chr&gt; | annotation &lt;chr&gt; | geneChr &lt;int&gt; | geneStart &lt;int&gt; | geneEnd &lt;int&gt; | geneLength &lt;int&gt; | geneStrand &lt;int&gt; | geneId &lt;chr&gt; | distanceToTSS &lt;dbl&gt; | flank_geneIds &lt;chr&gt; | geneSymbol &lt;chr&gt; | flank_geneSymbols &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     seqnames start end width strand interval_id annotation geneChr geneStart\n",
       "     geneEnd geneLength geneStrand geneId distanceToTSS flank_geneIds\n",
       "     geneSymbol flank_geneSymbols"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
