{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "624842cb-72b2-429a-94d8-b3d44dd9a2a5",
   "metadata": {},
   "source": [
    "# 2024-026 Peak Set Enrichment Analysis\n",
    "\n",
    "This notebook details processes for running PSEA on differential peaks obtained from 2024-026.\n",
    "\n",
    "**Expected Outputs**\n",
    " - CSV file of GSEA results\n",
    " - rds of raw ClusterProfiler objects"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f6528a6-4afa-4fcb-9315-4eadb2394ae8",
   "metadata": {},
   "source": [
    "### Initialize Environment\n",
    "\n",
    "Import the necessary packages and output package versions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b04df92b-f81b-41d6-b074-c66c8ef86e29",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Import necessary packages\n",
    "library(ggplot2)\n",
    "library(clusterProfiler)\n",
    "library(DOSE)\n",
    "library(scales)\n",
    "library(viridis)\n",
    "library(dplyr)\n",
    "\n",
    "# # Load data\n",
    "# working_dir <- \"/home/dalbao/2024-026-Tcf7ATAC/01_nfcore_241228\"\n",
    "\n",
    "# # Move to working directory\n",
    "# setwd(working_dir)\n",
    "\n",
    "# Load data\n",
    "exp_dir     <- \"/home/dalbao/2024-026-Tcf7ATAC\"\n",
    "nfcore_dir  <- \"01_nfcore_241228\"\n",
    "nb_dir      <- \"02_bp_notebooks\"\n",
    "prefix      <- \"DAnST-PSEA_noShrink_lfc\"\n",
    "\n",
    "# Set input DAPS (csv file)\n",
    "input <- \"250610-bp-DAnST-DAPS_noShrink-Unannotated.csv\"\n",
    "# Add folder structure to input\n",
    "input <- paste(nb_dir, \"daps\", input, sep = \"/\")\n",
    "\n",
    "# Peak Sets File Pattern\n",
    "ps_pattern <- \"noShrink.+peakSets.csv$\"              # Files with this pattern\n",
    "\n",
    "# nf-core ATAC-seq pipeline calls peaks from two different libraries:\n",
    "library     <- \"merged_library\"\n",
    "# If library == merged_library, lib_sh is mLb\n",
    "# else if it is == merged_replicates, lib_sh is mRp\n",
    "lib_sh      <- ifelse(library == \"merged_library\", \"mLb\", \"mRp\")\n",
    "\n",
    "# nf-core ATAC-seq pipeline calls peaks in two types, broad_peak and narrow_peak:\n",
    "peak_type   <- \"broad_peak\"\n",
    "# If peak_type == broad_peak, peak_sh is bp\n",
    "peak_sh     <- ifelse(peak_type == \"broad_peak\", \"bp\", \"np\")\n",
    "\n",
    "# Add peak type and date to prefix\n",
    "prefix <- paste0(format(Sys.Date(), \"%y%m%d\"), \"-\", peak_sh, \"-\", prefix)\n",
    "\n",
    "# Move to working directory\n",
    "setwd(exp_dir)\n",
    "\n",
    "# Toggle automatic data saving\n",
    "savedata <- TRUE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba171d41-40a8-4236-891e-c809e894df14",
   "metadata": {},
   "source": [
    "**Load gene sets and examine:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5794205d-85ed-4939-8646-ffdf42fbd722",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Loop and load all CSV files that end with peak_sets.csv in the directory.\n",
    "# List all CSV files that end with \"peak_sets.csv\" in the peak_sets directory\n",
    "peak_files <- list.files(path = paste(nb_dir, \"peak_sets\", sep = \"/\"), pattern = ps_pattern, full.names = TRUE)\n",
    "\n",
    "print(peak_files)\n",
    "\n",
    "# Initialize an empty list to hold data\n",
    "sigs_list <- list()\n",
    "\n",
    "# Loop through each file and read the data\n",
    "for (file in peak_files) {\n",
    "    sigs_list[[file]] <- read.csv(file)\n",
    "}\n",
    "\n",
    "# Combine all data frames in the list into one data frame\n",
    "sigs <- do.call(rbind, sigs_list)\n",
    "\n",
    "\n",
    "# Check peak list\n",
    "table(sigs$ps_name)\n",
    "# Show top 5 rows:\n",
    "head(sigs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b473c8be-2112-4773-a5ae-e1cea4768039",
   "metadata": {},
   "source": [
    "**Load dataset and examine:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07e3aa8c-d5bc-4cbc-9fd6-ef6f22fef840",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Load degs\n",
    "degs <- read.csv(input)\n",
    "\n",
    "# Reorder to make similar to scDatasets\n",
    "degs <- degs[, c(1,4,7,6,8,9)]\n",
    "# Rename\n",
    "colnames(degs) <- c(\"group\", \"names\", \"statistic\", \"scores\", \"pvals\", \"pvals_adj\")\n",
    "\n",
    "# Show top 5 rows\n",
    "head(degs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcf12a09-609c-416f-b30e-43c0e732f513",
   "metadata": {},
   "source": [
    "### Gene Set Enrichment Analysis\n",
    "\n",
    "Perform GSEA analyses per cluster and save these analyses to a list. Moreover, prepare degs to have an ordered list suitable for clusterProfiler; order genes from highest to lowest scores. First create a function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e9d072a-9f0e-4567-a9c4-d5cc59c4e86e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Convert peak sets to gene sets terminology\n",
    "colnames(sigs) <- c(\"gs_name\", \"gene_symbol\")\n",
    "\n",
    "# Define a function that takes the entire degs list, a cluster of interest, and gene signatures then returns a gseaResult object.\n",
    "performGSEA <- function(degs, cluster, sigs, seed = TRUE, set_seed = 42){\n",
    "    # Extract cluster of interest\n",
    "    cluster <- degs[degs$group == cluster, ]\n",
    "    # Extract Z scores\n",
    "    cluster.genes <- cluster$scores\n",
    "    # Name Z scores by gene\n",
    "    names(cluster.genes) <- cluster$names\n",
    "    # Sort genes by decreasing score (needed for clusterProfiler)\n",
    "    cluster.genes <- sort(cluster.genes, decreasing = TRUE)\n",
    "\n",
    "    # Perform GSEA\n",
    "    if(seed){\n",
    "        set.seed(42) # For computational reproducibility\n",
    "    }\n",
    "    gsea <- GSEA(   cluster.genes,\n",
    "                    TERM2GENE=sigs,\n",
    "                    nPerm = 10000,\n",
    "                    minGSSize = 5,\n",
    "                    maxGSSize = 20000,\n",
    "                    pvalueCutoff = 1,\n",
    "                    by=\"DOSE\",\n",
    "                    seed = seed)\n",
    "\n",
    "    return(gsea) # Return gseaResult object\n",
    "} # End of fucntion"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c10fca3-d6be-45cf-a89a-67d99bb22eff",
   "metadata": {},
   "source": [
    "Now that a function is defined, loop through all clusters within degs df:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baf0a2ab-8fee-4fcc-922b-43812e56c385",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Define a list to hold gseaResults\n",
    "enrichment_results = list()\n",
    "\n",
    "# Define an empty dataframe\n",
    "results_df <- data.frame()\n",
    "\n",
    "# Loop over all clusters in degs df\n",
    "for (cluster in unique(degs$group)){\n",
    "    # Assign gseaResult from performGSEA() to a specific item in enrichment_results\n",
    "    enrichment_results[cluster] <- performGSEA(degs = degs, cluster = cluster, sigs = sigs, seed = TRUE, set_seed = 42)\n",
    "\n",
    "    # Extract dataframe of results\n",
    "    df <- enrichment_results[[cluster]]@result\n",
    "    # Label cluster for the extracted dataframe\n",
    "    df$cluster <- cluster\n",
    "    # Put cluster column as the firstmost column\n",
    "    df <- df[, c(12, 1:11)]\n",
    "\n",
    "    # Rowbind dataframes to combine all clusters into one\n",
    "    # Conditional, if results_df is empty, assign current df:\n",
    "    if(dim(results_df)[1] == 0){\n",
    "        results_df <- df\n",
    "        print(\"Start\")\n",
    "    }else{ # If data exists, rowbind\n",
    "        results_df <- rbind.data.frame(results_df, df)\n",
    "    }\n",
    "\n",
    "    # Show progress\n",
    "    print(cluster)\n",
    "\n",
    "    rm(df, cluster) # Cleanup\n",
    "    gc(full = TRUE)\n",
    "}\n",
    "# End of loop"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ad59895-1f62-4366-b785-369f1e48c9b8",
   "metadata": {},
   "source": [
    "**Save CSV of GSEA results and the raw ClusterProfiler objects:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd52dff0-e237-4ac4-a1b1-053d695a3409",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Use BH p-adj\n",
    "results_df$BH <- p.adjust(results_df$pvalue, method = \"BH\")\n",
    "\n",
    "\n",
    "if(savedata){\n",
    "    save_dir <- paste(nb_dir, \"psea\", sep = \"/\")\n",
    "\n",
    "    # Save list containing gseaResult objects to RDS\n",
    "    saveRDS(enrichment_results, paste(save_dir, \"/\", prefix, \".rds\", sep = \"\"))\n",
    "\n",
    "    # Save list containing gseaResult objects to RDS\n",
    "    write.csv(results_df, paste(save_dir, \"/\", prefix, \".csv\", sep = \"\"), row.names = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce2ad238-72ae-4125-a405-e51bb91abf75",
   "metadata": {},
   "source": [
    "**Save the results of the differential expression analysis in a CSV file and the modified adata file:**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed6ae5fa-89c9-41b0-9e0e-83d25857e965",
   "metadata": {},
   "source": [
    "### Visualize GSEA Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1345cf7a-ecf8-4b2d-83ba-e288b104b761",
   "metadata": {},
   "source": [
    "**Pre-process data to order clusters accordingly and filter out non-significant results:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0816421-b894-4b25-8ce7-5089f6f2a92a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Calculate negative logarithm of q-value\n",
    "results_df$log_qvalue <- -log10(results_df$BH)\n",
    "\n",
    "# Filter significant q-values\n",
    "significant_df <- subset(results_df, BH < 0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4169309-6717-416c-8bb0-043b58440b66",
   "metadata": {},
   "source": [
    "**Define a function to plot data:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f09b5f30-ebf3-4d17-928a-a51c8d60897b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "plotGSEAbubble <- function(df, name){\n",
    "    # Define color scale\n",
    "    color_scale <- viridis::cividis(3)  # Generate a color scale with 2 colors\n",
    "\n",
    "    # Ensure limits are symmetric around 0\n",
    "    NES_range <- range(df$NES, na.rm = TRUE)\n",
    "    max_abs_NES <- max(abs(NES_range))\n",
    "    limits <- c(-max_abs_NES, max_abs_NES)\n",
    "\n",
    "    # Create bubble plot\n",
    "    p <- ggplot(df, aes(x = cluster, y = ID)) +\n",
    "      geom_point(aes(color = NES, size = log_qvalue)) +\n",
    "      scale_color_gradientn(\n",
    "        colors = color_scale,\n",
    "        limits = limits,\n",
    "        guide = guide_colorbar(title = \"NES\")\n",
    "      ) +\n",
    "      scale_size_continuous(\n",
    "        breaks = seq(0, max(df$log_qvalue), by = 0.5),\n",
    "        labels = seq(0, max(df$log_qvalue), by = 0.5),\n",
    "        guide = guide_legend(title = \"-log10(Q-value)\")\n",
    "      ) +\n",
    "      labs(\n",
    "        title = paste(\"Gene Set Enrichment\", name),  # Set the plot title\n",
    "        x = \"Cluster\",  # Label for the X-axis\n",
    "        y = \"ID\"  # Label for the Y-axis\n",
    "      ) +\n",
    "      theme_minimal() +  # Set the plot theme to minimal\n",
    "      theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate and align X-axis labels\n",
    "    return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4124bec8-49fe-4634-8c42-1ba37ff7d132",
   "metadata": {},
   "source": [
    "**Plot GSEAs by comparison:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1412502-dbc9-4014-b478-49d38f2fac90",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "subset_vs_naive   <- significant_df[!(grepl(\"_vs_\", significant_df$cluster)), ]\n",
    "print(plotGSEAbubble(subset_vs_naive, \"Subset vs Naive\"))\n",
    "shRunx3_vs_shCD19 <- significant_df[((grepl(\"shRunx3_vs\", significant_df$cluster)) & !(grepl(\"D5\", significant_df$cluster))), ]\n",
    "print(plotGSEAbubble(shRunx3_vs_shCD19, \"shRunx3 vs shCD19\"))\n",
    "intersubset       <- significant_df[(grepl(\"_vs_\", significant_df$cluster)) & (!grepl(\"shRunx3_vs_|D5\", significant_df$cluster)), ]\n",
    "print(plotGSEAbubble(intersubset, \"Intersubset Comparisons\"))\n",
    "intersubset       <- significant_df[(grepl(\"D5\", significant_df$cluster)), ]\n",
    "print(plotGSEAbubble(intersubset, \"Day 5 Comparisons\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
